src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java

```diff
@@ -408,92 +407,135 @@ public int register(NoGood noGood) {
-						return emptyList();
-					}
-				}
-			}
+			instances = singletonList(new Instance(substitute.getTerms()));
 		}
 
 		// substituted atom contains variables
 		if (currentLiteral.isNegated()) {
-			throw oops("Current atom should be positive at this point but is not");
+			if (laxGrounderHeuristic) {
+				return pushBackAndBindNextAtomInRule(groundingOrder, orderPosition, originalTolerance, remainingTolerance, partialSubstitution, currentAssignment);
+			} else {
+				throw oops("Current atom should be positive at this point but is not");
+			}
+		}
+
+		if (instances == null) {
+			instances = getInstancesForSubstitute(substitute, partialSubstitution);
+		}
+
+		if (laxGrounderHeuristic && instances.isEmpty()) {
+			// we have reached a point where we have to terminate binding,
+			// but it might be possible that a different grounding order would allow us to continue binding
+			// under the presence of a lax grounder heuristic
+			return pushBackAndBindNextAtomInRule(groundingOrder, orderPosition, originalTolerance, remainingTolerance, partialSubstitution, currentAssignment);
 		}
 
-		IndexedInstanceStorage storage = workingMemory.get(currentAtom.getPredicate(), true);
+		return createBindings(groundingOrder, orderPosition, originalTolerance, remainingTolerance, partialSubstitution, currentAssignment, instances, substitute);
+	}
+
+	private boolean shallPushBackFixedInterpretationLiteral(FixedInterpretationLiteral substitutedLiteral) {
+		return !(substitutedLiteral.isGround() ||
+				(substitutedLiteral instanceof ComparisonLiteral && ((ComparisonLiteral)substitutedLiteral).isLeftOrRightAssigning()) ||
+				(substitutedLiteral instanceof IntervalLiteral && substitutedLiteral.getTerms().get(0).isGround()) ||
+				(substitutedLiteral instanceof ExternalLiteral));
+	}
+
+	private Collection<Instance> getInstancesForSubstitute(Atom substitute, Substitution partialSubstitution) {
 		Collection<Instance> instances;
+		IndexedInstanceStorage storage = workingMemory.get(substitute.getPredicate(), true);
 		if (partialSubstitution.isEmpty()) {
-			if (currentLiteral.isGround()) {
-				instances = singletonList(new Instance(currentLiteral.getTerms()));
-			} else {
-				// No variables are bound, but first atom in the body became recently true, consider all instances now.
-				instances = storage.getAllInstances();
-			}
+			// No variables are bound, but first atom in the body became recently true, consider all instances now.
+			instances = storage.getAllInstances();
 		} else {
 			instances = storage.getInstancesFromPartiallyGroundAtom(substitute);
 		}
+		return instances;
+	}
 
-		ArrayList<Substitution> generatedSubstitutions = new ArrayList<>();
+	private BindingResult createBindings(RuleGroundingOrder groundingOrder, int orderPosition, int originalTolerance, int remainingTolerance, Substitution partialSubstitution, Assignment currentAssignment, Collection<Instance> instances, Atom substitute) {
```

Comment:
Please add comments/javadoc explaining what this method is intended to do (or why it was refactored).

Reply:
OK